src/librustc/session/filesearch.rs

```diff
@@ -64,14 +64,16 @@ impl<'a> FileSearch<'a> {
     pub fn search<F>(&self, mut pick: F)
         where F: FnMut(&Path, PathKind) -> FileMatch
     {
-        self.for_each_lib_search_path(|lib_search_path, kind| {
+        self.for_each_lib_search_path(&mut |lib_search_path: &Path, kind| {
             debug!("searching {}", lib_search_path.display());
-            let files = match fs::read_dir(lib_search_path) {
-                Ok(files) => files,
-                Err(..) => return,
-            };
-            let files = files.filter_map(|p| p.ok().map(|s| s.path()))
-                             .collect::<Vec<_>>();
+
+            // Get the pre-gathered `PathBuf`s for the files within
+            // `lib_search_path`. This always succeeds because
+            // `self.dir_files_map` is initialized when the `Session` is
+            // created to contain an entry for every directory that might be
+            // consulted here.
+            let files = self.dir_files_map.get(lib_search_path).unwrap();
```

Comment:
We shouldn't need to do this, whatever `for_each_lib_search_path` iterates over should contain the values of `dir_files_map` alongside the paths.

Reply:
`target_filesearch` and `host_filesearch` have overlap in the directories they consider. (In fact, on the examples I looked at, they were identical, presumably because I wasn't cross-compiling.) I can't see how to avoid this lookup without duplicating `PathBuf`s. I think having a separate map that covers both `target_filesearch` and `host_filesearch` is reasonable.

Reply:
I *really* don't want a map here, given how it's not actually needed.

If we had a type for "files in a directory" then we could have one value of that type per search path plus one for each of target and host in the sysroot (with one of them being `Option` in case no cross-compilation is happening).